### ISSUE TITLE
Simple concurrency safety breach.

### ISSUE THREAD

**Issue by CrystDragon** [swiftlang/swift#76929]
### Description

The following code compiles without warning.
```
nonisolated  func f() async -> Int {
    var value = 0
    let task1 = Task { @MainActor in
        for _ in 0..<100000 {
            value += 1
        }
    }
    for _ in 0..<100000 {
        value += 1
    }
    await _ = task1.value
    return value
}

print(await f())
```

But it cannot produce 200000 because of the apparent race on `value`.

Also, for the following modified code, the compiler will complain "Pattern that the region based isolation checker does not understand how to check".

```
nonisolated func f() async -> Int {
    var value = 0
    let task1 = Task { @MainActor in
        for _ in 0..<100000 {
            value += 1
        }
    }
    let task2 = Task { @MainActor in
        for _ in 0..<100000 {
            value += 1
        }
    }
    await _ = task1.value
    await _ = task2.value
    return value
}

```

### Reproduction

See above.

### Expected behavior

- There should be warning for the first example.
- Solve "Pattern that the region based isolation checker does not understand how to check" for the second example.

### Environment

- Apple Swift version 6.0 (swiftlang-6.0.0.9.10 clang-1600.0.26.2)
- Target: arm64-apple-macosx14.0
- Version 16.0 (16A242d)

**Comment by NSFatalError** [comment]
This issue seems to be related to the [problem I've described on Swift Forums](https://forums.swift.org/t/calling-non-isolated-async-method-from-a-child-task-inheriting-global-actor-allows-for-a-race-condition-in-swift-6/75332)

**Comment by CrystDragon** [comment]
Just tested this using the latest release/6.1 snapshot. There's still no diagnostics.
Isn't this one a crucial regression? @hborla

**Comment by CrystDragon** [comment]
Status update after tested against the release/6.1 nightly build.

For the second code snippet, now the compiler can emit a error saying "Sending 'value' risks causing data races"

But for the first code snippet, the compiler still fails to produce any error.
Not that if compiled using Swift 5.9, there will be an error.
![in Swift 5.9](https://github.com/user-attachments/assets/968887db-2f58-46ac-b714-0d12c5c63474).

@ktoso Forgive my ignorance, but don't you agree this a particular naive bad case? 
Are there any underlying complexities that make this one so hard to address?

**Comment by CrystDragon** [comment]
I would argue that in most Swift Concurrency cases, from a developer's view, "a false positive" could be a better option compared to "a false negative".
Personally I'd rather choose to be limited while free-of-error than to be uncertain of correctness.

**Comment by hborla** [comment]
Yes, this is a data-race safety hole that needs to be fixed. It's a bug in region based isolation.

**Comment by CrystDragon** [comment]
The nightly build can now correctly emit diagnostics for the 2 code snippets in OP. Good work!

**Comment by CrystDragon** [comment]
Just confirmed this is fixed in Swift 6.2 ~~
